### PR TITLE
Merge pull request #390 from gandrewstone/release

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

**Comment by gandrewstone** [comment]
this PR is basically what was already reviewed on the "release" branch, except for the last checkin which resolves an issue with locking that does not exist on "release".  Reviewers, please look at that checkin (aa16977)

File: src/txdb.cpp [hunk 1]
```diff
@@ -100,20 +100,31 @@ bool CBlockTreeDB::ReadLastBlockFile(int &nFile) {
...
+    {
+      cs_main.lock();
+      stats.hashBlock = GetBestBlock();
+      stats.nHeight = mapBlockIndex.find(stats.hashBlock)->second->nHeight;
+    }
+    catch (...)
+    {
+        cs_main.unlock();
+        throw;
+    }
```
Comment (ptschip) [inline src/txdb.cpp]: you don't have do this...cs_utxo is all you need.  Threads can only access the utxo if they have the cs_utxo lock...cs_main no longer has anything to do with accessing the utxo.  In the past, yes, everything was locked broadly, or assumed to be locked by cs_main, but that is no longer the case.   I would just leave it as it was,   LOCK(cs_main) in that small scope above and then let it go out of scope before grabbing the cs_utxo.

Comment (ptschip) [reply src/txdb.cpp]: Ok, i see what you're trying to do...i think you're right there...   The only thing I would change is to use ENTER_CRITICAL_SECTION and LEAVE_CRITICAL_SECTION because then the lockstack is getting updated correctly, using cs_main.lock() doesn't do that.

**Review by ptschip — COMMENTED** [review]
@gandrewstone  I don't see the fix for the :get_xthin...where we log a message and return if we can't read the block from disk?  or was that merged already?   Never, mind, it's early, i didn't load the diff...ok, it's fine.

File: contrib/gitian-descriptors/gitian-win.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ---
-name: "BitcoinUnlimited-win-0.12"
+name: "BitcoinUnlimited-win-1.0"
```
Comment (deadalnix) [inline contrib/gitian-descriptors/gitian-win.yml:2]: Version change in their own PR.

File: src/miner.cpp [hunk 3]
```diff
@@ -143,7 +143,7 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& s
       CBlockHeader h;
       nBlockSize += h.GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION);
     }
-    assert(nBlockSize == 80);  // BU always 80 bytes
+    assert(nBlockSize == 80);  // BU block header is always 80 bytes
```
Comment (deadalnix) [inline src/miner.cpp:146]: Putting comment on their own line make diff/history/blame clearer has changing the comment doesn't appear as changing the code.

File: src/rpc/server.cpp [hunk 4]
```diff
@@ -459,7 +459,7 @@ void RPCRegisterTimerInterface(RPCTimerInterface *iface)
 void RPCUnregisterTimerInterface(RPCTimerInterface *iface)
 {
     std::vector<RPCTimerInterface*>::iterator i = std::find(timerInterfaces.begin(), timerInterfaces.end(), iface);
-    assert(i != timerInterfaces.end());
+    DbgAssert(i != timerInterfaces.end(), return);  // already removed, so ignore the problem in production
```
Comment (deadalnix) [inline src/rpc/server.cpp:462]: Comment on its own line.

File: src/txdb.cpp [hunk 5]
```diff
@@ -100,15 +100,31 @@ bool CBlockTreeDB::ReadLastBlockFile(int &nFile) {
...
     CHashWriter ss(SER_GETHASH, PROTOCOL_VERSION);
     CAmount nTotalAmount = 0;
+
+    // I want to lock cs_utxo before unlocking cd_main so that a new block arrival
+    // does not have a moment to update the utxo set between our grabbing the
+    // hash and the height, and our utxo tally.
+    // But there is not reason to keep cs_main locked while we look at the utxo
+    try 
+    {
+      cs_main.lock();
```
Comment (deadalnix) [inline src/txdb.cpp]: Use RAII.

File: src/unlimited.cpp [hunk 6]
```diff
@@ -218,12 +217,13 @@ void HandleExpeditedRequest(CDataStream& vRecv,CNode* pfrom)
...
 	      {
 		LogPrint("blk", "Starting expedited blocks to peer %s (%d).\n", pfrom->addrName.c_str(),pfrom->id);
+                // find an empty array location
 		std::vector<CNode*>::iterator it = std::find(xpeditedBlk.begin(), xpeditedBlk.end(),((CNode*)NULL));
 		if (it != xpeditedBlk.end())
 		  *it = pfrom;
 		else
 		  xpeditedBlk.push_back(pfrom);
-		pfrom->AddRef();
+		pfrom->AddRef();  // add a reference because we have added this pointer into the expedited array
```
Comment (deadalnix) [inline src/unlimited.cpp:226]: Comment on its own line.

File: src/unlimited.cpp [hunk 7]
```diff
@@ -289,35 +289,35 @@ void HandleExpeditedBlock(CDataStream& vRecv,CNode* pfrom)
...
-
-      BlockMap::iterator mapEntry = mapBlockIndex.find(thinBlock.header.GetHash());
+      BlockMap::iterator mapEntry = mapBlockIndex.find(blkHash);
       CBlockIndex *blkidx = NULL;
       unsigned int status = 0;
       if (mapEntry != mapBlockIndex.end())
 	{
 	  blkidx = mapEntry->second;
-	  status = blkidx->nStatus;
+	  if (blkidx) status = blkidx->nStatus;
```
Comment (deadalnix) [inline src/unlimited.cpp:301]: I don't think it is expected normal operations to store a nullptr in the map. Unless this is the case, this should at least log, and probably bail because something went horribly wrong.

File: src/unlimited.cpp [hunk 8]
```diff
@@ -506,7 +508,7 @@ UniValue expedited(const UniValue& params, bool fHelp)
     std::vector<CNode*>::iterator elem = std::find(xpeditedBlkUp.begin(), xpeditedBlkUp.end(),node); 
     if ((flags & EXPEDITED_BLOCKS)&&(flags & EXPEDITED_STOP))
       {
-	if (elem == xpeditedBlkUp.end()) xpeditedBlkUp.erase(elem);
+	if (elem != xpeditedBlkUp.end()) xpeditedBlkUp.erase(elem);
```
Comment (deadalnix) [inline src/unlimited.cpp:511]: This is still lacking a test. The fact that the wrong condition could have lived for such a long time is testimony that a test was needed in the first place, and still is.

File: src/unlimited.cpp [hunk 9]
```diff
@@ -1482,23 +1484,27 @@ bool CheckAndRequestExpeditedBlocks(CNode* pfrom)
...
+              pos2 = strPeerIP.rfind("]:");
+              // Handle both ipv4 and ipv6 cases
+              if (pos1 <= 0 && pos2 <= 0) 
+                strListeningPeerIP = strPeerIP + ':' + boost::lexical_cast<std::string>(pfrom->addrFromPort);
+              else if (pos1 > 0)
+                strListeningPeerIP = strPeerIP.substr(0, pos1) + ':' + boost::lexical_cast<std::string>(pfrom->addrFromPort);
+              else
+                strListeningPeerIP = strPeerIP.substr(0, pos2) + ':' + boost::lexical_cast<std::string>(pfrom->addrFromPort);
+            }
+          else strListeningPeerIP = strPeerIP;
```
Comment (deadalnix) [inline src/unlimited.cpp:1507]: This should still be done in 2 steps: moving the code around and then do the changes.

File: src/main.cpp [hunk 10]
```diff
@@ -481,7 +481,7 @@ int64_t GetBlockTimeout(int64_t nTime, int nValidatedQueuedBefore, const Consens
 /** Check whether the last unknown block a peer advertised is not yet known. */
 void ProcessBlockAvailability(NodeId nodeid) {
     CNodeState *state = State(nodeid);
-    assert(state != NULL);
+    DbgAssert(state != NULL, return);  // node already destructed, nothing to do in production mode
```
Comment (deadalnix) [inline src/main.cpp:484]: Comment on its own line.

File: src/main.cpp [hunk 11]
```diff
@@ -496,7 +496,7 @@ void ProcessBlockAvailability(NodeId nodeid) {
 /** Update tracking information about which blocks a peer is assumed to have. */
 void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
     CNodeState *state = State(nodeid);
-    assert(state != NULL);
+    DbgAssert(state != NULL, return);  // node already destructed, nothing to do in production mode
```
Comment (deadalnix) [inline src/main.cpp:499]: Comment on its own line.

File: src/main.cpp [hunk 12]
```diff
@@ -2035,8 +2035,18 @@ bool CScriptCheck::operator()() {
...
 {
     LOCK(cs_main);
-    CBlockIndex* pindexPrev = mapBlockIndex.find(inputs.GetBestBlock())->second;
-    return pindexPrev->nHeight + 1;
+    BlockMap::iterator i = mapBlockIndex.find(inputs.GetBestBlock());
+    if (i != mapBlockIndex.end())
+    {
+        CBlockIndex* pindexPrev = i->second;
+        if (pindexPrev)
+            return pindexPrev->nHeight + 1;
```
Comment (deadalnix) [inline src/main.cpp:2043]: Apple engineers say: "use braces".

File: src/main.cpp [hunk 13]
```diff
@@ -2035,8 +2035,18 @@ bool CScriptCheck::operator()() {
 int GetSpendHeight(const CCoinsViewCache& inputs)
 {
     LOCK(cs_main);
-    CBlockIndex* pindexPrev = mapBlockIndex.find(inputs.GetBestBlock())->second;
-    return pindexPrev->nHeight + 1;
+    BlockMap::iterator i = mapBlockIndex.find(inputs.GetBestBlock());
+    if (i != mapBlockIndex.end())
```
Comment (deadalnix) [inline src/main.cpp:2039]: You should inverse that condition. Throw early if the block don't exist and avoid nesting all the logic.

File: src/main.cpp [hunk 14]
```diff
@@ -5088,55 +5105,62 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
...
+                                // Thus, the protocol spec specified allows for us to provide duplicate txn here,
+                                // however we MUST always provide at least what the remote peer needs
+                                typedef std::pair<unsigned int, uint256> PairType;
+                                BOOST_FOREACH(PairType& pair, merkleBlock.vMatchedTxn)
+                                {
+                                    pfrom->txsSent += 1;
+                                    pfrom->PushMessage(NetMsgType::TX, block.vtx[pair.first]);
+                                }
+                            }
+                            // else
```
Comment (deadalnix) [inline src/main.cpp:5149]: It looks like looking from the diff that this whole code was rewritten, but it ain't thee case. This makes it impossible to spot what actually changed. please split actual code change and moving code around in different PR. this is unreviewable in this form.

File: src/main.cpp [hunk 15]
```diff
@@ -5724,8 +5768,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                         // Probably non-standard or insufficient fee/priority
                         LogPrint("mempool", "   removed orphan tx %s\n", orphanHash.ToString());
                         vEraseQueue.push_back(orphanHash);
-                        assert(recentRejects);
-                        recentRejects->insert(orphanHash);
+                        if (recentRejects) recentRejects->insert(orphanHash);  // should always be true
```
Comment (deadalnix) [inline src/main.cpp:5771]: Braces + comment on its own line.

File: src/main.cpp [hunk 16]
```diff
@@ -5747,8 +5790,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             if (nEvicted > 0)
                 LogPrint("mempool", "mapOrphan overflow, removed %u tx\n", nEvicted);
         } else {
-            assert(recentRejects);
-            recentRejects->insert(tx.GetHash());
+            if (recentRejects) recentRejects->insert(tx.GetHash());  // should always be true
```
Comment (deadalnix) [inline src/main.cpp:5793]: Comment on its own line.

File: src/main.cpp [hunk 17]
```diff
@@ -5519,22 +5553,33 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
         {
             Misbehaving(pfrom->GetId(), 20);
             return error("message getdata size() = %u", vInv.size());
         }
 
+        // Validate that INVs are a valid type
+        for (unsigned int nInv = 0; nInv < vInv.size(); nInv++)
+        {
+            const CInv &inv = vInv[nInv];
+            if (!((inv.type == MSG_TX) || (inv.type == MSG_BLOCK) || (inv.type == MSG_FILTERED_BLOCK) || (inv.type == MSG_THINBLOCK) || (inv.type == MSG_XTHINBLOCK)))
```
Comment (deadalnix) [inline src/main.cpp:5574]: There is still a need for a wrapper for these conditions. Like

    inv.isTx() || inv.isSomeBlockType()

Or whatever. This makes it clear what is included and excluded, plus you can have truth table in tests and it'll avoid random code breaking when a new inv type is added.

File: src/main.cpp [hunk 18]
```diff
@@ -5919,23 +5962,37 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         CBloomFilter filterMemPool;
         CInv inv;
         vRecv >> inv >> filterMemPool;
+        if (!((inv.type == MSG_XTHINBLOCK)||(inv.type == MSG_THINBLOCK)))
```
Comment (deadalnix) [inline src/main.cpp:5965]: (inv.type != MSG_XTHINBLOCK) && (inv.type != MSG_THINBLOCK)

But even better:

    !inv.isSomeKindOfBlock()

File: src/main.cpp [hunk 19]
```diff
@@ -5944,7 +6001,12 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
       }
     else if (strCommand == NetMsgType::XPEDITEDBLK)  // BU
       {
-	HandleExpeditedBlock(vRecv,pfrom);
+	if (!HandleExpeditedBlock(vRecv,pfrom))
+        {
```
Comment (deadalnix) [inline src/main.cpp:6005]: I guess I should ask for Egyptian if I want Allman to happen here. I won't ask for Allman, because that'd get me Egyptian and now that most of the code is moving toward Allman...

File: src/main.cpp [hunk 20]
```diff
@@ -6100,6 +6159,21 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         int count=0;
         size_t i;
+        if (pfrom->xThinBlockHashes.size() != pfrom->thinBlock.vtx.size())  // Because the next loop assumes this
+          {
```
Comment (deadalnix) [inline src/main.cpp:6163]: Egyptian braces.

File: src/main.cpp [hunk 21]
```diff
@@ -6100,6 +6159,21 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         int count=0;
         size_t i;
+        if (pfrom->xThinBlockHashes.size() != pfrom->thinBlock.vtx.size())  // Because the next loop assumes this
```
Comment (deadalnix) [inline src/main.cpp:6162]: No comment on the same line.

File: src/main.cpp [hunk 22]
```diff
@@ -6704,7 +6781,7 @@ bool SendMessages(CNode* pto)
                 // headers that aren't on chainActive, give up.
                 BOOST_FOREACH(const uint256 &hash, pto->vBlockHashesToAnnounce) {
                     BlockMap::iterator mi = mapBlockIndex.find(hash);
-                    assert(mi != mapBlockIndex.end());
+                    if (mi == mapBlockIndex.end()) continue;  // BU skip blocks that we don't know about.  was: assert(mi != mapBlockIndex.end());
```
Comment (deadalnix) [inline src/main.cpp:6784]: comment on a different line.

**Comment by deadalnix** [comment]
Still no test, for pretty much any of the changes. Still a kitchen sink PR.

**Review by deadalnix — CHANGES_REQUESTED** [review]
Still no tests, still a kitchen sink PR.

**Comment by ptschip** [comment]
@gandrewstone you also need a LEAVE_CRITICAL... on line 116

**Comment by gandrewstone** [comment]
@ptschip thanks, fixed

**Review by ptschip — APPROVED** [review]

**Comment by ptschip** [comment]
@gandrewstone it all looks good to me now...
